### PR TITLE
ARROW-9412: [C++] Add non-bundled dependencies to INTERFACE_LINK_LIBRARIES of static libarrow

### PR DESCRIPTION
**Pull request by tobim**
This adds support for using the `arrow_static` CMake target with system static libraries.

This adds the following CMake options:

  * `ARROW_DEPENDENCY_USE_SHARED` (`ON` by default): The default value of `ARROW_*_USE_SHARED` except `ARROW_THRIFT_USE_SHARED` on Windows.
  * `ARROW_BZ2_USE_SHARED`: Use shared library of bzip2.
  * `ARROW_LZ4_USE_SHARED`: Use shared library of LZ4.
  * `ARROW_THRIFT_USE_SHARED` (`OFF` by default on Windows): Use shared library of libthrift. This is `OFF` by
     default on Windows because it seems that Thrift's DLL support isn't well for not.
  * `ARROW_UTF8PROC_USE_SHARED`: Use shared library of utf8proc.
  * `ARROW_SNAPPY_USE_SHARED`: Use shared library of Snappy.
  * `ARROW_ZSTD_USE_SHARED`: Use shared library of ZStandard.

Note that users must specify `ARROW_*_USE_SHARED=OFF` explicitly when they want to use static library. This is a backward
incompatible change for some cases. For example, users can use shared library or static library without specifying any options
before this change. Because old `FindLz4.cmake` find both shared library and static library.

Note that upstream `SnappyConfig.cmake` and `zstdConfig.cmake` are never used with this change. We always use our
`Find*.cmake` for them to process `ARROW_*_USE_SHARED`.

Note that the `ZSTD_SOURCE` CMake option is renamed to `zstd_SOURCE`.

Some dependencies don't support static linking with `ARROW_*_USE_SHARED=OFF` yes. For example, gRPC may ignore it.
If `gRPCConfig.cmake` is installed, our `FindgRPCAlt.cmake` isn't used. It means that `ARROW_GRPC_USE_SHARED=OFF`
is ignored. We can work on this as further tasks.

RE2 is only required by Gandiva but it's searched by `find_package(Arrow)` not `find_package(Gandiva)`. It should be fixed
by a follow-up task.

### CODE REVIEW

**Comment by kou** [comment]
Thanks.
I can push the branch.

I'm still working on this. Please wait for a while.

File: cpp/cmake_modules/FindLz4.cmake [hunk 1]
```diff
@@ -24,13 +24,14 @@ set(LZ4_STATIC_LIB_SUFFIX "${LZ4_MSVC_STATIC_LIB_SUFFIX}${CMAKE_STATIC_LIBRARY_S
 set(LZ4_LIB_NAMES lz4 liblz4)
 
 if(NOT ARROW_LZ4_USE_SHARED)
+  set(static_names_)
```
Comment (tobim) [inline cpp/cmake_modules/FindLz4.cmake]: This is not necessary.

File: cpp/src/arrow/ArrowConfig.cmake.in [hunk 2]
```diff
@@ -38,45 +38,42 @@ include("${CMAKE_CURRENT_LIST_DIR}/ArrowOptions.cmake")
...
-endif()
-
-
 # Load targets only once. If we load targets multiple times, CMake reports
 # already existent target error.
 if(NOT (TARGET arrow_shared OR TARGET arrow_static))
   include("${CMAKE_CURRENT_LIST_DIR}/ArrowTargets.cmake")
 
   if(TARGET arrow_static)
+    set(LIB_PATH_SUFFIXES
```
Comment (tobim) [inline cpp/src/arrow/ArrowConfig.cmake.in]: This change means that the static dependencies need to be resolved even if `arrow_static` is never used.

File: cpp/cmake_modules/Findzstd.cmake [hunk 3]
```diff
@@ -54,10 +54,11 @@ else()
   if(ZSTD_PC_FOUND)
     set(ZSTD_INCLUDE_DIR "${ZSTD_PC_INCLUDEDIR}")
 
-    if(NOT ARROW_ZSTD_USE_SHARED)
+    if(ARROW_ZSTD_USE_SHARED)
+      list(APPEND ZSTD_PC_LIBRARY_DIRS "${ZSTD_PC_LIBDIR}")
+    else()
```
Comment (tobim) [inline cpp/cmake_modules/Findzstd.cmake:63]: Some libzstd.pc versions are written in a way that only ${ZSTD_PC_STATIC_LIBDIR} contains a path. I started out with essentially this, but it didn't work during testing.

File: cpp/cmake_modules/FindLz4.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/FindLz4.cmake]: If someone defines `static_name_` before calling `FindLz4.cmake`, the `static_name_` value is used.
Is it wrong?

File: cpp/src/arrow/ArrowConfig.cmake.in (hunk 2 above)
Comment (kou) [reply cpp/src/arrow/ArrowConfig.cmake.in]: Yes. Is it strange?
Do you know a CMake convention when a CMake package provides both shard and static libraries?
Do users need to specify "we want to use static library" explicitly?

Should we use the `COMPONENTS` feature in `find_package`?

File: cpp/cmake_modules/Findzstd.cmake (hunk 3 above)
Comment (kou) [reply cpp/cmake_modules/Findzstd.cmake:63]: Oh...
Could you show one of the libzstd.pc versions environment? I want to try it.

File: cpp/cmake_modules/FindSnappy.cmake [hunk 4]
```diff
@@ -15,25 +15,33 @@
 # specific language governing permissions and limitations
 # under the License.
 
+set(SNAPPY_STATIC_LIB_SUFFIX "${SNAPPY_MSVC_STATIC_LIB_SUFFIX}${CMAKE_STATIC_LIBRARY_SUFFIX}")
```
Comment (kou) [inline cpp/cmake_modules/FindSnappy.cmake:18]: Is this needed?
It seems that Snappy doesn't add any suffix for MSVC: https://github.com/google/snappy/blob/master/CMakeLists.txt

File: cpp/cmake_modules/FindSnappy.cmake [hunk 5]
```diff
@@ -15,25 +15,33 @@
...
   find_path(Snappy_INCLUDE_DIR
             NAMES snappy.h
             PATHS ${Snappy_ROOT}
             NO_DEFAULT_PATH
-            PATH_SUFFIXES ${INCLUDE_PATH_SUFFIXES})
+            PATH_SUFFIXES ${ARROW_INCLUDE_PATH_SUFFIXES})
 else()
-  find_library(Snappy_LIB NAMES snappy)
-  find_path(Snappy_INCLUDE_DIR NAMES snappy.h PATH_SUFFIXES ${INCLUDE_PATH_SUFFIXES})
+  find_library(Snappy_LIB NAMES ${SNAPPY_LIB_NAMES} HINTS "${CMAKE_ROOT}/Modules/")
```
Comment (kou) [inline cpp/cmake_modules/FindSnappy.cmake:38]: Could you explain why the `HINTS` is needed?
I don't think that `libsnappy.so` exists in `${CMAKE_ROOT}/Modules/`.

File: cpp/cmake_modules/Findzstd.cmake (hunk 3 above)
Comment (kou) [reply cpp/cmake_modules/Findzstd.cmake:63]: It seems that `XXX_STATIC_LIBDIR` is never defined: https://github.com/Kitware/CMake/blob/master/Modules/FindPkgConfig.cmake#L566
`_pkg_get_variable` doesn't set `XXX_STATIC_YYY`: https://github.com/Kitware/CMake/blob/master/Modules/FindPkgConfig.cmake#L566
(`XXX_LIBDIR` is only defined.)

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly

**Comment by kou** [comment]
@tobim Most of changes are done. I'm working on the last check. Could you confirm the current code and the updated PR description? If the last check is done and addressed your feedback, I'll merge this.

Here are summary of my changes https://github.com/apache/arrow/compare/76b4b022179f94bdb5b77086acf253b6a7bcbbe9...3aed2131e18a7f9db00cd446fc887d78e777686a :

  * Export `ARROW_LIBRARY_PATH_SUFFIXES` and `ARROW_INCLUDE_PATH_SUFFIXES` to `ArrowConfig.cmake` instead of copying definitions of `LIB_PATH_SUFFIXES` and `INCLUDE_PATH_SUFFIXES`.
  * Rename `ARROW_USE_SHARED_SYSTEM_LIBS` to `ARROW_DEPENDENCY_USE_SHARED` because we already have `ARROW_DEPENDENCY_SOURCE` option that is the default value of `*_SOURCE`.
  * Don't use `ARROW_DEPENDENCY_USE_SHARED` as the default value of `ARROW_THRIFT_USE_SHARED_DEFAULT` on Windows because it seems that Thrift's DLL support isn't well for now.
  * Use more strict library names for LZ4, ZStandard, Snappy, Thrift and utf8proc. I confirmed library names of them on WIndows by provided binaries and build configuration files of them.
  * Use only package names (`ARROW_SYSTEM_DEPENDENCIES`) instead of `find_dependency(...)` (`ARROW_FIND_DEPENDENCY_LIST`) to provide `find_dependency()` list to `ArrowConfig.cmake` because it'll be simpler.
  * Use Docker Compose instead of raw Docker in `cpp/examples/minimal_build/` to manage multiple build configurations.
    * I split one `Dockerfile` that contains `*-dev` packages to `Dockerfile` with/without `*-dev` packages to show our build system uses bundled libraries for nonexistent dependencies by default.

The AppVeyor failure is unrelated: https://issues.apache.org/jira/browse/ARROW-9621

**Comment by kou** [comment]
@github-actions crossbow submit -g example -g linux

**Comment by kou** [comment]
@github-actions crossbow submit -g linux wheel-win-*

**Comment by kou** [comment]
@github-actions crossbow submit centos-* wheel-win-*

**Review by kou — APPROVED** [review]
+1

I'll merge this. If we find any problem, we can work on it as a follow-up task.

**Comment by tobim** [comment]
I just tested this overnight and luckily everything works for my needs.

**Comment by tobim** [comment]
@kou thank you for seeing this through!

File: cpp/cmake_modules/FindSnappy.cmake (hunk 4 above)
Comment (tobim) [reply cpp/cmake_modules/FindSnappy.cmake]: Thanks for double-checking.

File: cpp/cmake_modules/FindSnappy.cmake (hunk 5 above)
Comment (tobim) [reply cpp/cmake_modules/FindSnappy.cmake]: I don't remember why I added this. It was probably a conversion error.

**Comment by kou** [comment]
Thanks for testing this!

**Comment by wesm** [comment]
Thanks @tobim!

**Pull request closed without merging**